spec/versionedsubmitblockrequest.go

```diff
@@ -0,0 +1,531 @@
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Bellatrix.ExecutionPayload.GasLimit, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Capella.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Capella.ExecutionPayload.GasLimit, nil
+	default:
+		return 0, errors.New("unsupported version")
+	}
+}
+
+// GasUsed returns the prev randao of the payload
+func (v *VersionedSubmitBlockRequest) GasUsed() (uint64, error) {
+	if v == nil {
+		return 0, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Bellatrix.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Bellatrix.ExecutionPayload.GasUsed, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Capella.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Capella.ExecutionPayload.GasUsed, nil
+	default:
+		return 0, errors.New("unsupported version")
+	}
+}
+
+// BlockNumber returns the block number of the payload.
+func (v *VersionedSubmitBlockRequest) BlockNumber() (uint64, error) {
+	if v == nil {
+		return 0, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Bellatrix.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Bellatrix.ExecutionPayload.BlockNumber, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Capella.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Capella.ExecutionPayload.BlockNumber, nil
+	default:
+		return 0, errors.New("unsupported version")
+	}
+}
+
+// Timestamp returns the timestamp of the payload.
+func (v *VersionedSubmitBlockRequest) Timestamp() (uint64, error) {
+	if v == nil {
+		return 0, errors.New("nil struct")
+	}
+	switch v.Version {
+	case consensusspec.DataVersionBellatrix:
+		if v.Bellatrix == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Bellatrix.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Bellatrix.ExecutionPayload.Timestamp, nil
+	case consensusspec.DataVersionCapella:
+		if v.Capella == nil {
+			return 0, errors.New("no data")
+		}
+		if v.Capella.ExecutionPayload == nil {
+			return 0, errors.New("no data execution payload")
+		}
+		return v.Capella.ExecutionPayload.Timestamp, nil
+	default:
+		return 0, errors.New("unsupported version")
+	}
+}
+
+// Transactions returns the transactions of the payload
+func (v *VersionedSubmitBlockRequest) Transactions() ([]consensusbellatrix.Transaction, error) {
```

Comment:
I'm a little concerned about this and `Withdrawals()` returning specific versions of their structs.  If there was a `consensusdeneb.Transaction` then the signature of this function would have to change.  But I'm not sure that the alternative of putting in place a versioned structure (or an interface) would be significantly better...

Reply:
I don't see transaction as raw `[]bytes` likely to change but there may be a case for withdrawals. mainly using the transaction field to log the number of transactions and calculate the withdrawals root. Can have functions that calculate those fields but would rather user call `HashTreeRoot()` on the structs.

Reply:
I think we'll have to go ahead as things are, and accept that the signature may change in future.